File: x-fly/src/main/java/com/xia/fly/mvp/nullobject/Defaults.java
package com.xia.fly.mvp.nullobject;

import java.util.HashMap;
import java.util.Map;

import static java.util.Collections.unmodifiableMap;

/**
 * @author xia
 * @date 2018/7/19.
 */
public final class Defaults {

    private static final Map<Class<?>, Object> DEFAULTS =
            unmodifiableMap(new HashMap<Class<?>, Object>() {
                {
                    put(Boolean.TYPE, false);
                    put(Byte.TYPE, (byte) 0);
                    put(Character.TYPE, '\000');
                    put(Double.TYPE, 0.0d);
                    put(Float.TYPE, 0.0f);
                    put(Integer.TYPE, 0);
                    put(Long.TYPE, 0L);
                    put(Short.TYPE, (short) 0);
                }
            });

    @SuppressWarnings("unchecked")
    public static <T> T defaultValue(Class<T> type) {
        return (T) DEFAULTS.get(type);
    }
}

File: fly/src/main/java/com/xia/fly/mvp/nullobject/Defaults.kt
package com.xia.fly.mvp.nullobject

import java.util.*
import java.util.Collections.unmodifiableMap

/**
 * @author xia
 * @date 2018/7/19.
 */
@Suppress("UNCHECKED_CAST")
object Defaults {

    private val DEFAULTS = unmodifiableMap(object : HashMap<Class<*>, Any>() {
        init {
            put(java.lang.Boolean.TYPE, false)
            put(java.lang.Byte.TYPE, 0.toByte())
            put(Character.TYPE, '\u0000')
            put(java.lang.Double.TYPE, 0.0)
            put(java.lang.Float.TYPE, 0.0f)
            put(Integer.TYPE, 0)
            put(java.lang.Long.TYPE, 0L)
            put(java.lang.Short.TYPE, 0.toShort())
        }
    })

    @JvmStatic
    fun <T> defaultValue(type: Class<T>?): T? {
        type ?: return null
        return DEFAULTS[type] as? T
    }
}
